File: code/ex224.java
import java.util.*;

class Solution {
    public void doCalculation(Stack<String> num) {
        while (num.size() > 1 && num.peek() != "(") {
            int num2 = Integer.parseInt(num.pop());
            if (num.peek() == "(") {
                num.pop();
                if (!num.isEmpty() && num.peek() == "-") {
                    num.pop();
                    num.push(String.valueOf(-num2));
                }
                else    num.push(String.valueOf(num2));
                break;
            }
            int num1 = Integer.parseInt(num.pop());
            
            num.push(String.valueOf(num1 + num2));
        }
    }
    
    
    public int calculate(String s) {
        Stack<String> num = new Stack<>();

        boolean sign = false;
        for (int i = 0; i < s.length(); i++) {
            char c = s.charAt(i);
            if (c == ')')    doCalculation(num);
            else if (c == '-') {
                if (s.charAt(i + 1) == '(')    num.push("-");
                else    sign = true;
            }
            else if (c == '(')    num.add("(");
            else if (c >= 48 && c <= 57) {
                String n = "";
                while (i < s.length() && s.charAt(i) >= 48 && s.charAt(i) <= 57) {
                    n += Character.toString(s.charAt(i));
                    i++;
                }
                
                if (sign)    num.add("-" + n);
                else    num.add(n);
                
                sign = false;    i--;
            }
        }
        
        doCalculation(num);
        
        return Integer.parseInt(num.pop());
    }
}

File: code/ex224.cpp
class Solution {
    private int i = 0;
    public int calculate(String s) {
        int n = s.length();
        int res = 0, temp = 0, sign = 1;
        while (i < n) {
            char c = s.charAt(i);

            if (c == ' ') {
                i++;
                continue;
            } else if (Character.isDigit(c)) {
                temp = temp * 10 + (c - '0');
                i++;
            } else {
                res += sign * temp;
                temp = 0;

                if (c == '+') {
                    sign = 1;
                    i++;
                } else if (c == '-') {
                    sign = -1;
                    i++;
                } else if (c == '(') {
                    i++;
                    int nxt = calculate(s);
                    if (nxt == Integer.MIN_VALUE)    return nxt;
                    else    res += sign * nxt;
                } else if (c == ')') {
                    i++;
                    return res;
                }
            }
        }
        if (temp != 0)    res += temp * sign;

        return res;
    }
}